### ISSUE TITLE
[Bug] Is `init_module()` called on old version of the contract when upgrading

### ISSUE THREAD

**Issue by rafales** [aptos-labs/aptos-core#3573]
# 🐛 Bug

It seems that upgrading a contract via `aptos move publish` calls `init_module()` on previous version of a contract

## To reproduce

* I narrowed down example to this:
```
module initbug::InitBug {
    use std::string::{Self, String};
    use std::signer::address_of;

    struct InitMessage has key, store {
        message: String,
        counter: u64
    }

    fun init_module(account: &signer) acquires InitMessage {
        let addr = address_of(account);
        let msg = string::utf8(b"hello one");

        if(exists<InitMessage>(addr)) {
            let im = borrow_global_mut<InitMessage>(addr);
            im.message = msg;
            im.counter = im.counter + 1;
        } else {
            let im = InitMessage { message: msg, counter: 1 };
            move_to(account, im);
        }
    }
}
```

Changing msg to `"hello two"` and re-publishing increases the counter but the message stored is still `"hello one"`.

My original issue was with a coin implementation - `init_module()` was calling `coin::initialize()` - subsequent attempts to publish the contract resulted in `0x80002` error which meant attempt to re-initialize a coin. Even leaving `init_module()` empty or getting rid of it didn't help.

## Expected Behavior

I would expect that either:
* `init_module()` gets called once and that's it
* or `init_module()` gets called with new implementation, not old one

## System information

I am publishing modules to current devnet. Aptos CLI version 0.3.1. MacOS.

**Comment by rafales** [comment]
@areshand I see that you posted the PR which introduced `init_module()`. Could you weight in?

**Comment by rafales** [comment]
Here is a minimal example: https://github.com/rafales/aptos-contracts-annotated/tree/main/03-init-bug

**Comment by wrwg** [comment]
Thanks for tracking this down.

**Comment by wrwg** [comment]
In general, if you upgrade a module, the new code is not available until the next transaction is executed. (In previous versions of Aptos, it wasn't even available until the next block,  but this should be fixed by now.) Can you shed some light on how you upgrade and how you measure which init is called? (That is what are the transactions executed here.) Also which version of Aptos are you using?

**Comment by rafales** [comment]
@wrwg I'm deploying to devnet with Aptos CLI 0.3.1.

I simply re-publish code with `aptos move publish`.

First time I ran into the bug I was calling `coin::initialize()` and on re-publish it was aborted with code which translated to "this coin is already initialized". So I removed `init_module()`'s body, but still got same error on re-publish (as if it was running old code).

So I decided to test it and created the code you see above. To reproduce:

* Publish module with `aptos move publish`
* Re-publish again to make sure it works with 'aptos move publish'
* Now if you check the resource the contact created - counter should be set to `2` and message to `"hello one"`
* Change `        let msg = string::utf8(b"hello one");` line so message is `"hello two"`
* Re-publish with `aptos move publish`
* Now - if new update was called then counter should be set to `3` and `message` to `"hello two"`
* But what I am seeing is counter being set to `3` and `message` to `"hello one"`
* This means that `init_module()` was ran as expected but the old version was executed
